File: Programação orientada a objetos/POO-241 - Trab 2 - Pedro_Quedevez/Ocorrencia.cpp
#include <iostream>
#include <string>
#include <vector>
#include "StringUtils.hpp"
#include "Ocorrencia.hpp"
#include "Tokenizer.hpp"
#include "DateUtils.hpp"
#include "Excecao.hpp"

using namespace cpp_util;
using namespace std;

Ocorrencia::Ocorrencia(string csvline)
{
    Tokenizer tokens(csvline, ';');
    vector<string> campos = tokens.remaining();

    for (int i = 0; i < campos.size(); i++)
    {
        campos[i] = trim(campos[i]);
    }

    try
    {

        if (codigosUtilizados.find(campos[0]) == codigosUtilizados.end())
        {
            throw Excecao("Código de docente não definido usado na ocorrência: " + campos[0] + ":" + campos[1] + ".");
        }
        this->setCodigo(campos[0]);
        this->setEvento(campos[1]);
        this->setDataInicio(campos[2]);
        this->setDataFim(campos[3]);
    }
    catch (Excecao e)
    {
        throw e;
    }
}

string Ocorrencia::getCodigo()
{
    return this->codigo;
}

string Ocorrencia::getEvento()
{
    return this->evento;
}

time_t Ocorrencia::getDataInicio()
{
    return this->dataInicio;
}

time_t Ocorrencia::getDataFim()
{
    return this->dataFim;
}

void Ocorrencia::setCodigo(string s)
{
    this->codigo = s;
}

void Ocorrencia::setEvento(string s)
{
    this->evento = s;
}

void Ocorrencia::setDataInicio(time_t t)
{
    this->dataInicio = t;
}

void Ocorrencia::setDataFim(time_t t)
{
    this->dataFim = t;
}

void Ocorrencia::setDataInicio(string s)
{
    this->dataInicio = cpp_util::parseDate(trim(s), cpp_util::DATE_FORMAT_PT_BR_SHORT);
}

void Ocorrencia::setDataFim(string s)
{
    this->dataFim = cpp_util::parseDate(trim(s), cpp_util::DATE_FORMAT_PT_BR_SHORT);
}

ostream &operator<<(ostream &c, Ocorrencia &o)
{
    c << o.getCodigo() << " " << o.getEvento() << " " << cpp_util::formatDate(o.getDataInicio(), cpp_util::DATE_FORMAT_PT_BR_SHORT) << " " << cpp_util::formatDate(o.getDataFim(), cpp_util::DATE_FORMAT_PT_BR_SHORT);
    return c;
}

File: Programação orientada a objetos/POO 241 - Trab 1 - Pedro Quedevez/Ocorrencia.java
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;

public class Ocorrencia {
    private Long codigo;
    private String evento;
    private LocalDate dataInicio;
    private LocalDate dataFim;

    public Ocorrencia(String csvline) throws excecao {
        String[] campos = csvline.split(";");
        try {
            if (campos.length == 4) {
                if (!Docente.codigosCadastrados.contains(Long.parseLong(campos[0]))) {
                    throw new excecao(
                            "Código de docente não definido usado na ocorrência: " + campos[0] + ":" + campos[1]
                                    + ".");
                }
                this.codigo = Long.parseLong(campos[0].trim());
                this.evento = campos[1].trim();
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
                this.dataInicio = LocalDate.parse(campos[2].trim(), formatter);
                this.dataFim = LocalDate.parse(campos[3].trim(), formatter);
            }
        } catch (java.lang.NumberFormatException e) {
            throw new excecao("Erro de formatacao");
        }
    }

    // auxilia na verificação de eventos de recadastramento;
    public static String EventoRecadastramento(ArrayList<Ocorrencia> ocorrencias, int ano) {
        for (Ocorrencia ocorrencia : ocorrencias) {
            String evento = ocorrencia.getEvento().trim();
            evento = evento.replaceAll(" ", "");

            if (evento.equals("Coordenador")) {
                return "Coordenador";
            } else if (evento.equals("LicençaMaternidade")) {
                if (ocorrencia.getDataFim().getYear() == (ano - 1)
                        || (ocorrencia.getDataInicio().getYear() == ano
                                && ocorrencia.getDataInicio().getDayOfMonth() == 1
                                && ocorrencia.getDataInicio().getMonthValue() == 1)) {
                    return "Licença Maternidade";
                }
            } else if (evento.equals("BolsistaCNPq")) {
                LocalDate dataComparar = LocalDate.of(ano, 1, 1);
                if ((ocorrencia.getDataInicio().isBefore(dataComparar) && ocorrencia.getDataFim().isAfter(dataComparar))
                        || (ocorrencia.getDataInicio().getYear() == ano
                                && ocorrencia.getDataInicio().getDayOfMonth() == 1
                                && ocorrencia.getDataInicio().getMonthValue() == 1)
                        || (ocorrencia.getDataFim().getYear() == ano
                                && ocorrencia.getDataFim().getDayOfMonth() == 1
                                && ocorrencia.getDataFim().getMonthValue() == 1))
                    return "Bolsista CNPq";
            }
        }
        return "Nao";
    }

    @Override
    public String toString() {
        return codigo + ";" + evento + ";" + dataInicio + ";" + dataFim;
    }

    public Long getCodigo() {
        return codigo;
    }

    public void setCodigo(Long codigo) {
        this.codigo = codigo;
    }

    public String getEvento() {
        return evento;
    }

    public void setEvento(String evento) {
        this.evento = evento;
    }

    public LocalDate getDataInicio() {
        return dataInicio;
    }

    public void setDataInicio(LocalDate dataInicio) {
        this.dataInicio = dataInicio;
    }

    public LocalDate getDataFim() {
        return dataFim;
    }

    public void setDataFim(LocalDate dataFim) {
        this.dataFim = dataFim;
    }

}
